### PR TITLE
Store offset periods as UTC

### PR DESCRIPTION
**Pull request by robertbastian**
Metazone periods are already stored as UTC. This gets us closer to joining the two data structs.

It's also less ambiguous than local data.

#6238

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -328,3 +328,6 @@ lto = true
 [profile.bench-memory]
 inherits = "bench"
 debug = true
+
+[workspace.metadata.cargo-semver-checks.lints]
+type_method_marked_deprecated.level = "allow"
```
Comment (Manishearth) [inline Cargo.toml]: thought: we should probably not be erroring on anything that CSC has as a "minor version bump" lint; since our patch releases are truly _patch_ releases (we release them with a patch on a release branch, not with an updated `main`)

File: components/time/src/zone/mod.rs [hunk 2]
```diff
@@ -385,9 +385,24 @@ impl TimeZoneInfo<models::Base> {
...
+    ///
+    /// Also see [`Self::with_zone_name_timestamp`].
     pub fn at_date_time_iso(self, date_time: DateTime<Iso>) -> TimeZoneInfo<models::AtTime> {
-        Self::with_zone_name_timestamp(self, ZoneNameTimestamp::from_date_time_iso(date_time))
+        Self::with_zone_name_timestamp(
+            self,
+            ZoneNameTimestamp::from_zoned_date_time_iso(crate::ZonedDateTime {
+                date: date_time.date,
+                time: date_time.time,
+                // If we don't have an offset, interpret as UTC. This is incorrect during a couple of hours in 50 years
```
Comment (Manishearth) [inline components/time/src/zone/mod.rs]: question: why? is this just the metazone changes thing?

Comment (robertbastian) [reply components/time/src/zone/mod.rs]: why what? have you looked at it commit by commit?

Comment (Manishearth) [reply components/time/src/zone/mod.rs]: Why is this incorrect during a couple of hours in 50 years?

No; I have not (I only started reviewing), but I think this comment should be self contained

File: components/datetime/src/format/input.rs [hunk 3]
```diff
@@ -95,8 +95,8 @@ impl DateTimeInputUnchecked {
     }
 
     /// Sets the local time for time zone name resolution.
-    pub fn set_time_zone_name_timestamp(&mut self, local_time: ZoneNameTimestamp) {
-        self.zone_name_timestamp = Some(local_time);
+    pub fn set_time_zone_name_timestamp(&mut self, timestamp: ZoneNameTimestamp) {
+        self.zone_name_timestamp = Some(timestamp);
     }
```
Comment (sffc) [inline components/datetime/src/format/input.rs:100]: Thought: Having a pre-existing wrapper type ZoneNameTimestamp makes this sort of changes easier

File: Cargo.toml (hunk 1 above)
Comment (sffc) [reply Cargo.toml]: Suggestion: Can you clarify what this does? Add a comment, please. I do not see much documentation for what effect `type_method_marked_deprecated` has.

File: components/time/Cargo.toml [hunk 4]
```diff
@@ -47,3 +47,6 @@ serde = ["dep:serde", "zerovec/serde", "zerotrie/serde", "tinystr/serde", "icu_p
 datagen = ["serde", "dep:databake", "zerovec/databake", "zerotrie/databake", "tinystr/databake", "icu_provider/export", "icu_locale_core/databake"]
 compiled_data = ["dep:icu_time_data", "icu_calendar/compiled_data", "icu_provider/baked"]
 alloc = ["tinystr/alloc", "zerotrie/alloc", "serde?/alloc"]
+
+[package.metadata.cargo-semver-checks.lints]
+workspace = true
```
Comment (sffc) [inline components/time/Cargo.toml:52]: Question: What does this do?

File: components/time/src/zone/mod.rs [hunk 5]
```diff
@@ -385,9 +385,26 @@ impl TimeZoneInfo<models::Base> {
         }
     }
 
-    /// Sets the [`ZoneNameTimestamp`] to the given local datetime.
+    /// Sets the [`ZoneNameTimestamp`] to the given datetime.
```
Comment (sffc) [inline components/time/src/zone/mod.rs:388]: Praise: Good docs and comments here

File: components/time/src/zone/zone_name_timestamp.rs [hunk 6]
```diff
@@ -211,6 +236,19 @@ impl ZoneNameTimestamp {
...
+    // to ZNT.
+    // This saturates at 0 and u32::MAX, which is correct for the use case of looking up
+    // in the offset periods data struct
+    pub(crate) fn adjust_zone(self, from: UtcOffset, to: UtcOffset) -> Self {
+        let diff = (to.to_seconds() - from.to_seconds()) / 60 / 15;
+        Self(if diff < 0 {
+            self.0.saturating_sub(diff.unsigned_abs())
+        } else {
+            self.0.saturating_add(diff.unsigned_abs())
+        })
```
Comment (sffc) [inline components/time/src/zone/zone_name_timestamp.rs:249]: Praise: Nice overflow-safe math

File: ffi/capi/src/timezone.rs [hunk 7]
```diff
@@ -151,25 +152,73 @@ pub mod ffi {
...
+                            zone: self.offset.unwrap_or(icu_time::zone::UtcOffset::zero()),
+                        },
+                    ),
+                ),
+                ..*self
+            })
+        }
+
+        /// Sets the timestamp at which to interpret the time zone
+        /// for display name lookup.
```
Comment (sffc) [inline ffi/capi/src/timezone.rs]:
```suggestion
        /// Sets the timestamp, in milliseconds since Unix epoch, at which to interpret the time zone
        /// for display name lookup.
```

File: ffi/capi/src/timezone.rs [hunk 8]
```diff
@@ -151,25 +152,73 @@ pub mod ffi {
...
+            FnInStruct,
+            hidden
+        )]
+        #[diplomat::rust_link(
+            icu::time::zone::ZoneNameTimestamp::far_in_future,
+            FnInStruct,
+            hidden
+        )] // documented
+        #[diplomat::rust_link(icu::time::zone::ZoneNameTimestamp::far_in_past, FnInStruct, hidden)] // documented
+        pub fn at_timestamp(&self, timestamp: i64) -> Box<Self> {
```
Comment (sffc) [inline ffi/capi/src/timezone.rs]: Suggestion:
```suggestion
        #[diplomat::rust_link(icu::time::TimeZoneInfo::with_zone_name_timestamp, FnInStruct)]
        #[diplomat::rust_link(icu::time::zone::ZoneNameTimestamp::from_zoned_date_time_iso, FnInStruct, compact)]
        #[diplomat::rust_link(icu::time::zone::ZoneNameTimestamp, Struct, compact)]
        pub fn at_timestamp(&self, timestamp: i64) -> Box<Self> {
```

File: ffi/capi/src/variant_offset.rs [hunk 9]
```diff
@@ -148,11 +148,57 @@ pub mod ffi {
             icu::time::zone::VariantOffsetsCalculatorBorrowed::compute_offsets_from_time_zone_and_name_timestamp,
             FnInStruct
         )]
+        #[diplomat::rust_link(
+            icu::time::zone::ZoneNameTimestamp::from_zoned_date_time_iso,
+            FnInStruct,
+            hidden
+        )]
```
Comment (sffc) [inline ffi/capi/src/variant_offset.rs]:
```suggestion
```

File: ffi/capi/src/variant_offset.rs [hunk 10]
```diff
@@ -148,11 +148,57 @@ pub mod ffi {
...
         pub fn compute_offsets_from_time_zone_and_date_time(
```
Comment (sffc) [inline ffi/capi/src/variant_offset.rs:152]: Suggestion: make an issue or TODO to remove this function in 3.0?

File: ffi/capi/src/variant_offset.rs [hunk 11]
```diff
@@ -148,11 +148,57 @@ pub mod ffi {
...
+                .compute_offsets_from_time_zone_and_name_timestamp(
+                    time_zone.0,
+                    icu_time::zone::ZoneNameTimestamp::from_zoned_date_time_iso(
+                        icu_time::ZonedDateTime {
+                            date: utc_date.0,
+                            time: utc_time.0,
+                            zone: icu_time::zone::UtcOffset::zero(),
+                        },
+                    ),
+                )?;
```
Comment (sffc) [inline ffi/capi/src/variant_offset.rs]: Suggestion: use the deprecated API here (with a lint suppression) so that when the deprecated API is deleted, we also remember to delete this function

File: ffi/capi/src/zoned_date_formatter.rs [hunk 12]
```diff
@@ -591,9 +591,10 @@ pub mod ffi {
                 input.set_time_zone_name_timestamp(zone_name_timestamp);
             }
             else {
-                input.set_time_zone_name_timestamp(icu_time::zone::ZoneNameTimestamp::from_date_time_iso(icu_time::DateTime {
+                input.set_time_zone_name_timestamp(icu_time::zone::ZoneNameTimestamp::from_zoned_date_time_iso(icu_time::ZonedDateTime {
                     date: iso_date.0,
-                    time: icu_time::Time::noon()
+                    time: icu_time::Time::noon(),
+                    zone: zone.offset.unwrap_or(icu_time::zone::UtcOffset::zero()),
```
Comment (sffc) [inline ffi/capi/src/zoned_date_formatter.rs]: Suggestion: Would be nice if this fallback to UTC lived in the icu_time crate instead of in icu_capi

File: provider/source/src/cldr_serde/time_zones/meta_zones.rs [hunk 13]
```diff
@@ -116,15 +117,22 @@ fn deserialize_date<'de, D: serde::de::Deserializer<'de>>(
...
-    let date_time = DateTime::try_from_str(&timestamp, Iso)
+    let DateTime { date, time } = DateTime::try_from_str(&timestamp, Iso)
         .map_err(|_| D::Error::custom("Invalid metazone timestamp"))?;
 
-    Ok(Some(ZoneNameTimestamp::from_date_time_iso(date_time)))
+    Ok(Some(ZoneNameTimestamp::from_zoned_date_time_iso(
+        ZonedDateTime {
+            date,
+            time,
+            zone: UtcOffset::zero(),
```
Comment (sffc) [inline provider/source/src/cldr_serde/time_zones/meta_zones.rs:135]: Praise: Thanks for the bug fix here. Suggestion: Add a test for the behavior of a metazone transition near the boundary: the last ZoneNameTmestamp that is on the old name and the first ZoneNameTimestamp that is on the new name. Bonus points: back it up with a citation to a press release outside of CLDR.

File: provider/source/src/time_zones/mod.rs [hunk 14]
```diff
@@ -79,12 +81,69 @@ impl SourceDataProvider {
                     .supplemental
                     .meta_zones;
```
Comment (sffc) [inline provider/source/src/time_zones/mod.rs:83]: Note: I did a cursory review of this datagen code, but like most datagen code, I'm more interested in if it works and passes tests. I would rather rely on good test coverage than code review, because this stuff is complicated and I'm not sure my time-consuming detailed review would turn up more issues. Since it is just data, bugs are relatively easy to fix without breaking things.

**Review by sffc — APPROVED** [review]
Praise: Nice work! Thanks for doing this. Glad we're aligned that this is the right way to go.

File: Cargo.toml (hunk 1 above)
Comment (Manishearth) [reply Cargo.toml]: https://github.com/obi1kenobi/cargo-semver-checks/blob/2643e1e03da2d921271e881195a7a053b4a71094/src/lints/type_method_marked_deprecated.ron#L3
